File: apps/host/internal/server/webroot_test.go
package server

import (
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func fixtureWebRoot(t *testing.T) (*WebRoot, string) {
	t.Helper()
	directory := t.TempDir()
	if err := os.MkdirAll(filepath.Join(directory, "assets"), 0755); err != nil {
		t.Fatal(err)
	}
	for name, body := range map[string]string{
		"index.html":                 "<!doctype html><title>shell</title>",
		"assets/index-abc12345.js":   "export const a = 1;\n",
		"assets/index-abc12345.css":  ":root{color:red}\n",
		"favicon.ico":                "icon",
		"assets/font-abc12345.woff2": "font",
	} {
		if err := os.WriteFile(filepath.Join(directory, filepath.FromSlash(name)), []byte(body), 0644); err != nil {
			t.Fatal(err)
		}
	}
	web, err := OpenWebRoot(directory)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { web.Close() })
	return web, directory
}

func serveStatic(t *testing.T, web *WebRoot, method, path string) *http.Response {
	t.Helper()
	recorder := httptest.NewRecorder()
	web.Serve(recorder, httptest.NewRequest(method, path, nil))
	return recorder.Result()
}

func TestWebRootRequiresAnApplicationShell(t *testing.T) {
	if _, err := OpenWebRoot(t.TempDir()); err == nil {
		t.Fatal("a directory without index.html was accepted")
	}
	if _, err := OpenWebRoot(filepath.Join(t.TempDir(), "missing")); err == nil {
		t.Fatal("a missing directory was accepted")
	}
	if _, err := OpenWebRoot(""); err == nil {
		t.Fatal("an empty directory was accepted")
	}
}

func TestWebRootServesTheShellAndItsAssets(t *testing.T) {
	web, _ := fixtureWebRoot(t)
	for _, expectation := range []struct{ path, kind, contains string }{
		{"/", "text/html; charset=utf-8", "shell"},
		{"/workspaces/abc/deep/link", "text/html; charset=utf-8", "shell"},
		{"/assets/index-abc12345.js", "text/javascript; charset=utf-8", "export"},
		{"/assets/index-abc12345.css", "text/css; charset=utf-8", "color"},
		{"/assets/font-abc12345.woff2", "font/woff2", "font"},
	} {
		response := serveStatic(t, web, http.MethodGet, expectation.path)
		body, _ := io.ReadAll(response.Body)
		response.Body.Close()
		if response.StatusCode != 200 || response.Header.Get("Content-Type") != expectation.kind || !strings.Contains(string(body), expectation.contains) {
			t.Fatalf("%s answered %d %q %q", expectation.path, response.StatusCode, response.Header.Get("Content-Type"), body)
		}
		if !strings.Contains(response.Header.Get("Content-Security-Policy"), "frame-ancestors 'none'") {
			t.Fatalf("%s was served without a policy", expectation.path)
		}
	}
	// Hashed assets may be cached; the shell itself never is, so a redeployed
	// front end is picked up on the next load.
	if cache := serveStatic(t, web, http.MethodGet, "/assets/index-abc12345.js").Header.Get("Cache-Control"); !strings.Contains(cache, "immutable") {
		t.Fatalf("a hashed asset was not cacheable: %q", cache)
	}
	if cache := serveStatic(t, web, http.MethodGet, "/").Header.Get("Cache-Control"); strings.Contains(cache, "immutable") {
		t.Fatalf("the shell was made immutable: %q", cache)
	}
}

func TestWebRootRefusesTraversalAndMissingAssets(t *testing.T) {
	web, directory := fixtureWebRoot(t)
	secret := filepath.Join(filepath.Dir(directory), "secret.txt")
	if err := os.WriteFile(secret, []byte("private"), 0600); err != nil {
		t.Fatal(err)
	}
	for _, path := range []string{"/../secret.txt", "/assets/../../secret.txt", "/./../secret.txt"} {
		response := serveStatic(t, web, http.MethodGet, path)
		body, _ := io.ReadAll(response.Body)
		response.Body.Close()
		if strings.Contains(string(body), "private") {
			t.Fatalf("%s escaped the web root", path)
		}
	}
	// A missing hashed asset must never be answered with HTML: a stale client
	// would then try to execute the application shell as JavaScript.
	response := serveStatic(t, web, http.MethodGet, "/assets/gone-00000000.js")
	response.Body.Close()
	if response.StatusCode != 404 {
		t.Fatalf("a missing asset answered %d", response.StatusCode)
	}
	response = serveStatic(t, web, http.MethodPost, "/")
	response.Body.Close()
	if response.StatusCode != 405 {
		t.Fatalf("a POST to the shell answered %d", response.StatusCode)
	}
}

func TestWebRootDoesNotFollowSymlinksOutOfTheBundle(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("symlink creation needs a privilege this test does not require")
	}
	web, directory := fixtureWebRoot(t)
	secret := filepath.Join(filepath.Dir(directory), "escape.txt")
	if err := os.WriteFile(secret, []byte("private"), 0600); err != nil {
		t.Fatal(err)
	}
	if err := os.Symlink(secret, filepath.Join(directory, "assets", "escape-00000000.js")); err != nil {
		t.Fatal(err)
	}
	response := serveStatic(t, web, http.MethodGet, "/assets/escape-00000000.js")
	body, _ := io.ReadAll(response.Body)
	response.Body.Close()
	if strings.Contains(string(body), "private") {
		t.Fatal("a symlink inside the bundle escaped the web root")
	}
}

File: apps/server/src/web-root.test.ts
import { mkdirSync, symlinkSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { describe, expect, it } from "vitest";
import {
  cacheControlFor,
  contentTypeFor,
  hashedAsset,
  openWebRoot,
  resolveFile,
  resolveWithinRoot,
  staticHeaders,
} from "./web-root";
import { tempDir } from "../../desktop/src/core/testing/temp-dir";

function fixture(): { root: string; outside: string } {
  const root = tempDir("armadra-webroot-");
  const outside = tempDir("armadra-outside-");
  writeFileSync(join(outside, "secret.txt"), "不该被读到");
  writeFileSync(join(root, "index.html"), "<!doctype html>");
  mkdirSync(join(root, "assets"));
  writeFileSync(join(root, "assets", "app-D3fK9x2a.js"), "1");
  symlinkSync(join(outside, "secret.txt"), join(root, "link.txt"));
  symlinkSync(join(root, "assets", "app-D3fK9x2a.js"), join(root, "inside.js"));
  return { root, outside };
}

describe("静态托管", () => {
  it("字符串上的收敛：解码、塌缩、落在根外就没有下一步", () => {
    // `resolve` 而不是写死 `/srv/web`：Windows 上根会带盘符、分隔符是反斜杠，
    // 收敛逻辑本身是跨平台的，断言也该按平台的拼法来。
    const root = resolve("/srv/web");
    expect(resolveWithinRoot(root, "/index.html")).toBe(
      join(root, "index.html"),
    );
    expect(resolveWithinRoot(root, "/a//b/../index.html")).toBe(
      join(root, "a", "index.html"),
    );
    // 编码过的 `..` 与直接写的 `..` 塌缩成同一个答案，而它在根外。
    expect(resolveWithinRoot(root, "/%2e%2e%2f%2e%2e%2fetc")).toBeUndefined();
    expect(resolveWithinRoot(root, "/../../etc")).toBeUndefined();
    expect(resolveWithinRoot(root, "/index.html?v=1")).toBe(
      join(root, "index.html"),
    );
    expect(resolveWithinRoot(root, "/%ff")).toBeUndefined();
    expect(resolveWithinRoot(root, "/a\0b")).toBeUndefined();
  });

  it("realpath 上的复核挡住指向包外的符号链接", async () => {
    const { root } = fixture();
    const web = await openWebRoot(root);
    expect(await resolveFile(web, "/link.txt")).toBeUndefined();
    // 指向包内的链接照常服务。
    expect((await resolveFile(web, "/inside.js"))?.relative).toBe(
      join("assets", "app-D3fK9x2a.js"),
    );
  });

  it("单页回退只接住没有扩展名的路径", async () => {
    const { root } = fixture();
    const web = await openWebRoot(root);
    expect((await resolveFile(web, "/workspace/abc"))?.relative).toBe(
      "index.html",
    );
    expect(await resolveFile(web, "/assets/missing.js")).toBeUndefined();
  });

  it("没有 index.html 的目录不是 apps/web 的产物", async () => {
    const empty = tempDir("armadra-empty-");
    await expect(openWebRoot(empty)).rejects.toThrow(/index\.html/);
  });

  it("缓存按产物形状分档", () => {
    expect(hashedAsset("app-D3fK9x2a.js")).toBe(true);
    expect(hashedAsset("plain.js")).toBe(false);
    expect(cacheControlFor("index.html")).toBe("no-store");
    expect(cacheControlFor(join("assets", "app-D3fK9x2a.js"))).toBe(
      "public, max-age=31536000, immutable",
    );
    expect(cacheControlFor("favicon.ico")).toBe("no-cache");
  });

  it("类型与安全头", () => {
    expect(contentTypeFor("/a/b.html")).toBe("text/html; charset=utf-8");
    expect(contentTypeFor("/a/b.unknown")).toBe("application/octet-stream");
    const headers = staticHeaders();
    expect(headers["x-content-type-options"]).toBe("nosniff");
    expect(headers["content-security-policy"]).toContain("default-src 'self'");
    expect(headers["cache-control"]).toBe("no-store");
  });
});
